src/pessimisticPatternMatcherUtils.cc

```diff
@@ -32,16 +32,134 @@ namespace lsst {
+                ndarray::asEigenMatrix(src_delta_array[src_idx]).dot(ndarray::asEigenMatrix(src_ctr));
+
+        ndarray::Array<double, 1, 1> proj_src_delta =
+                copy(src_delta_array[src_idx] - proj_src_delta_dot * src_ctr);
+        double geom_dist_src =
+                sqrt(ndarray::asEigenMatrix(proj_src_delta).dot(ndarray::asEigenMatrix(proj_src_delta)) *
+                     proj_src_ctr_dist_sq);
+
+        /// Compute cosine and sine of the delta vector opening angle.
+        double cos_theta_src =
+                ndarray::asEigenMatrix(proj_src_delta).dot(ndarray::asEigenMatrix(proj_src_ctr_delta)) /
+                geom_dist_src;
+        Eigen::Vector3d cross_src = ndarray::asEigenMatrix(proj_src_delta)
+                                            .head<3>()
+                                            .cross(ndarray::asEigenMatrix(proj_src_ctr_delta).head<3>()) /
+                                    geom_dist_src;
+        double sin_theta_src = cross_src.dot(ndarray::asEigenMatrix(src_ctr));
+
+        /// Test the spokes and return the id of the reference object.
+        /// Return -1 if no match is found.
+        int ref_id =
+                check_spoke(cos_theta_src, sin_theta_src, ref_ctr, proj_ref_ctr_delta, proj_ref_ctr_dist_sq,
+                            candidate_range, ref_id_array, reference_array, src_sin_tol);
+        if (ref_id < 0) {
+            n_fail++;
+            continue;
+        }
+
+        /// Append the successful indices to our list. The src_idx needs
+        /// an extra iteration to skip the first and second source objects.
+        std::vector<size_t> tmpIds{ref_id, src_idx + 1};
+        output_spokes.push_back(tmpIds);
+        if (output_spokes.size() >= n_match - 2) {
+            return output_spokes;
+        }
+    }
+    return output_spokes;
+}
+
 int check_spoke(double cos_theta_src, double sin_theta_src, ndarray::Array<double, 1, 1> const& ref_ctr,
                 ndarray::Array<double, 1, 1> const& proj_ref_ctr_delta, double proj_ref_ctr_dist_sq,
-                ndarray::Array<long int, 1, 1> const& ref_dist_idx_array,
+                std::vector<size_t> const& candidate_range,
                 ndarray::Array<uint16_t, 1, 1> const& ref_id_array,
                 ndarray::Array<double, 2, 1> const& reference_array, double src_sin_tol) {
     // Loop over our candidate reference objects. ref_dist_idx_array is a view into ref_id_array
     // and is not the same length as ref_id_array.
-    for (auto idx = ref_dist_idx_array.begin(); idx != ref_dist_idx_array.end(); idx++) {
+    size_t midpoint = candidate_range[1];
+    for (size_t idx = 0; idx < candidate_range[2] - candidate_range[0]; idx++) {
         // Compute the delta vector from the pattern center.
-        unsigned int ref_id = ref_id_array[*idx];
+        if (idx % 2 == 0) {
+            midpoint = midpoint + idx;
+            ;
+        } else {
+            midpoint = midpoint - idx;
+        }
```

Comment:
I'd recommend factoring this logic into its own function, like you did in Python; it's much harder to figure out what these alternating indices do when they're embedded in a larger data access loop.

Reply:
Hmm... I'm a little hesitant doing this as it would mean modifying midpoint in place in the external function. I've change things around slightly now that I'm just returning the start and end value from find_candidate.

Reply:
I guess I assumed that the function could return a vector giving the correct access order, and `midpoint` would become an iteration variable over that vector instead of `idx` (which you seem to use exclusively for calculating `midpoint`).

Reply:
I suppose. I was thinking once all the functions that use `find_candidate` are in C++ i could write something where the `find_candidate` method returns and iterator that does what you describe. I'm hesitant to create a vector as I don't want to have to loop through this data more than once. Depending on the tolerance, the iterations these steps do can be pretty large.

Reply:
Fair enough -- I realize this is still work in progress.

That said, creating an iterator is a lot more work in C++ than in Python, so I was suggesting the vector of indices as a quick solution.

Reply:
Yeah, I figured. I think I'm happy enough with it for now but could add some extra comments.